### PR TITLE
Complete exercies chapter/13

### PR DESCRIPTION
**Pull request by nguyentrieutien1**
Comparison between build and new
- new is a method provided by Ruby for creating a new instance of a class.
- build is a method provided by ActiveRecord, the ORM (Object-Relational Mapping) library in Rails. It is used to create a new object with its associations automatically built
+ Associations in Rails
- acssociations in Rails define the relationships between models. They allow you to establish connections between different tables in the database.
+ Comparison between scope and class methods
- scope is a method in Rails that allows you to define reusable query conditions on a model
- class methods, on the other hand, are regular methods defined within a model class that can be called directly on the class itself
+ Why not to use default_scope
- can lead to unexpected behavior and make it harder to reason about the queries being executed. It can also cause performance issues, especially when combined with other scopes or when eager loading associations.
+ Delegate
- delegate is a method provided by ActiveSupport in Rails that allows you to delegate a method call from one object to another
- it is commonly used when you want to provide a convenient interface for accessing methods from associated objects without explicitly defining forwarding methods
by using delegate, you can avoid duplicating methods or exposing unnecessary implementation details.
+ Comparison between count, size, and length
- count is a method in Rails that returns the number of records that match a given condition
- size and length are similar methods that return the number of records in a collection or association
+ N+1 query
-  N refers to the number of main objects, and 1 refers to the initial query
- To fix N+1 queries, you can use eager loading techniques such as includes, preload, or eager_load to fetch the associated objects in the initial query, reducing the number of database hits.

### CODE REVIEW

**Comment by nguyentrieutien1** [comment]
ready

**Comment by nguyentrieutien1** [comment]
ready
image

File: app/controllers/microposts_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+class MicropostsController < ApplicationController
+  before_action :logged_in_user, only: %i(create destroy)
+  before_action :correct_user, only: :destroy
+
+  def new
+    @micropost = Micropost.new
+  end
+
+  def create
+    @micropost = @current_user.microposts.build micropost_params
```
Comment (vanvtt-0952) [inline app/controllers/microposts_controller.rb:12]: Hỏi câu: build và new khác nhau thế nào cu

File: app/javascript/custom/image_upload.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+document.addEventListener("turbo:load", function () {
+    document.addEventListener("change", function (event) {
+        let image_upload = document.querySelector('#micropost_image');
+        const size_in_megabytes = image_upload.files[0].size / 1024 / 1024;
+        if (size_in_megabytes > 5) {
+            alert("Maximum file size is 5MB. Please choose a smaller file.");
```
Comment (vanvtt-0952) [inline app/javascript/custom/image_upload.js]: I18n cho text này, em tìm hiểu vè I18n-js gem nhé

File: app/views/microposts/_micropost.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<li class="mb-5 d-flex col-4 m-auto" id="micropost-<%= micropost.id %>">
+  <div class=""><%= link_to gravatar_for(micropost.user, size: Settings["DIGIT_140"]), micropost.user %></div>
+  <div class="ms-5">
+    <p class="user"><%= link_to micropost.user.name, micropost.user %></p>
+    <p class="content"><%= micropost.content %></p>
+    <%= image_tag micropost.image if micropost.image.attached? %>
+    <p class="timestamp">Posted <%= time_ago_in_words micropost.created_at %>
```
Comment (vanvtt-0952) [inline app/views/microposts/_micropost.html.erb]: I18n nè :leu leu

File: app/views/microposts/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<% if logged_in? %>
+  <div class="row">
+    <section class="user_info col-4">
+      <%= render "shared/profile" %>
+    </section>
+    <section class="micropost_form col-8">
+      <%= render "form" %>
+    </section>
+  </div>
+<% else %>...
```
Comment (vanvtt-0952) [inline app/views/microposts/new.html.erb]: ủA ALOOOOO
... là sao cụ

File: app/views/shared/_feed.html.erb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<% if @microposts.any? %>
+  <ol class="microposts">
+    <%= render @microposts %>
+  </ol>
+  <%== pagy_bootstrap_nav @pagy %>
+<% end %>
```
Comment (vanvtt-0952) [inline app/views/shared/_feed.html.erb]:
```suggestion
  <ol class="microposts">
    <%= render(@microposts) || t "microposts_empty")%>
  </ol>
  <%== pagy_bootstrap_nav @pagy %>
```
tự tạo i18n `microposts_empty` đi

File: app/views/shared/_profile.html.erb [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<%= gravatar_for @current_user %>
+<div class="card-body">
+  <h5 class="card-title">
+    <%= @current_user["name"] %>
```
Comment (vanvtt-0952) [inline app/views/shared/_profile.html.erb]:
```suggestion
    <%= @current_user.name %>
```
ri dc ko cu

File: app/views/static_pages/home.html.erb [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
-    <%= link_to t("button.login"), "#", class: "btn btn-lg btn-primary" %>
+<% if logged_in? %>
+  <%= render "shared/feed" %>
+<% else %>
+  <div>
+    <div class="center jumbotron">
+      <h1>
+        <%= t("application.welcome") %>
+      </h1>
+      <h2 class="btn trix-button">
```
Comment (vanvtt-0952) [inline app/views/static_pages/home.html.erb]: I18nnnn

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -1,8 +1,14 @@
-<div class="card w-25">
-  <%= gravatar_for @user %>
-  <div class="card-body">
-    <h5 class="card-title">
-      <%= @user["name"] %>
-    </h5>
-  </div>
+<div class="row">
+	<div class="col-8  w-25">
```
Comment (vanvtt-0952) [inline app/views/users/show.html.erb]: indent cu cậu

File: app/controllers/microposts_controller.rb (hunk 1 above)
Comment (nguyentrieutien1) [reply app/controllers/microposts_controller.rb:12]: Theo em hiểu thì new chỉ tạo instance trên class đó, thằng build nó cũng tạo instance nhưng nó sẽ auto map relationship với thằng cha ( ví dụ parent.children.build(name: "children1") thì thằng children sẽ tự tạo pho ren ki cho parent với id đó  )

**Comment by nguyentrieutien1** [comment]
ready

File: app/controllers/microposts_controller.rb (hunk 1 above)
Comment (vanvtt-0952) [reply app/controllers/microposts_controller.rb:12]: chưa đủ nha, thử với rails c đi

File: app/javascript/locales/en.js [hunk 9]
```diff
@@ -0,0 +1,6 @@
+window.I18n = {
```
Comment (vanvtt-0952) [inline app/javascript/locales/en.js]: ủa có ri à =))
https://github.com/fnando/i18n-js

File: app/views/static_pages/home.html.erb [hunk 10]
```diff
@@ -1,11 +1,15 @@
...
-    <h2 class="btn trix-button">
-      <%= t("application.description", link: '<a href="http://www.railstutorial.org/">Ruby on Rails Tutorial</a>') %>
-    </h2>
-    <%= link_to t("button.login"), "#", class: "btn btn-lg btn-primary" %>
-  </div>
-</div>
+<% if logged_in? %>
+	<%= render "shared/feed" %>
+<% else %>
+	<div>
```
Comment (vanvtt-0952) [inline app/views/static_pages/home.html.erb:4]: file này sai indent nè, thấy đang dùng tab, em check lại nha

File: app/views/users/show.html.erb [hunk 11]
```diff
@@ -1,8 +1,14 @@
...
+		<%= render "shared/profile" %>
```
Comment (vanvtt-0952) [inline app/views/users/show.html.erb]: file này em đang dùng tab rồi, đổi thành 2 space nào

**Comment by vanvtt-0952** [comment]
merged
có comment nhỏ, fix rồi mợt
cái tội ko self review, bữa sao cứ 1 comment ~ 5k nhé, rẻ bèo như cá ngoài chợ

**Pull request merged**